File: 3121-count-the-number-of-special-characters-ii/3121. Count the Number of Special Characters II.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        res = 0
        dic = defaultdict(list)
        for i, letter in enumerate(word):
            dic[letter].append(i)
        visited = set()
        for i in range(len(word)):
            if word[i] not in visited and word[i].lower() == word[i] and word[i].upper() in dic and dic[word[i].upper()][0] > dic[word[i]][-1]: 
                visited.add(word[i])
                res+=1
        return res

File: 3121-count-the-number-of-special-characters-ii/3121. Count the Number of Special Characters II.java
class Solution {
    public int numberOfSpecialChars(String word) {
        HashMap<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < word.length(); i++) {
            if (word.charAt(i) == Character.toUpperCase(word.charAt(i)) && map.containsKey(word.charAt(i))) continue;
            map.put(word.charAt(i), i);
        }
        int res = 0;
        HashSet<Character> seen = new HashSet<Character>();
        for (int i = 0; i < word.length(); i++) {
            char upper = Character.toUpperCase(word.charAt(i));
            char lower = Character.toLowerCase(word.charAt(i));

            if (map.containsKey(lower) && map.containsKey(upper) && map.get(lower) < map.get(upper)) {
                seen.add(lower);
            }
        }
        return seen.size();

    }
}